### PR TITLE
Allow you to tab through modal text box

### LINKED ISSUE #16004: Navigating through a form with atom-text-editor as inputs with tab key
I opened an issue in the GitHub package repo (https://github.com/atom/github/issues/1234) with a gif demonstrating the problem. 

I'm seeing the same behavior when I use `<atom-text-editor>`s as input elements. The focus moves from the input to the next button.

### PR DESCRIPTION
**Pull request by jasonrudolph**
Fixes #16004

Supersedes #16127

---

This pull request builds on @itsmichaelwang's changes in #16127 and attempts to resolve the [possible drawbacks noted in that pull request](https://github.com/atom/atom/pull/16127#issue-151324878). (Please see #16127 for background on the overall approach.)

@itsmichaelwang: Thank you for getting this started! :bow:

### Demo

The demo below illustrates:

- <kbd>tab</kbd> to focus next element
- <kbd>shift</kbd>+<kbd>tab</kbd> to focus previous element
- <kbd>escape</kbd> to dismiss modal

demo

### TODO

- [x] Allow tab navigation from one `atom-text-editor` element to another ordered by tabIndex
- [x] Allow tab navigation from an `atom-text-editor` element to another element ordered by tabIndex
- [x] Get sanity check from maintainers regarding unintended side effects this change might have
- [x] Add a test for specifying a custom tab index on an `atom-text-editor` element
- [x] Document and execute [verification process](https://github.com/atom/atom/pull/16484)

### Verification Process

- [x] Verify that you can tab through text fields and buttons in `github:clone` dialog. (See demo gif.)
- [x] Verify that you can still tab through fields in the find-and-replace dialog.
- [x] Verify that <kbd>tab</kbd> still inserts a tab (or spaces, depending on your preferences) when editing a buffer
- [x] When a line is selected in a buffer, verify that <kbd>tab</kbd> still indents the line and that <kbd>shift</kbd>+<kbd>tab</kbd> still outdents the line.

### CODE REVIEW

**Comment by jasonrudolph** [comment]
> - [ ] Get sanity check from maintainers regarding unintended side effects this change might have

@nathansobo @as-cii: This pull request changes the tabIndex code introduced in https://github.com/atom/atom/commit/c52d66377fbd90087fbe3d970e82ecd9c020fcbb. Given your work on that commit, would you mind reviewing this pull request? I'm wondering if you know of any potential side effects that I should be aware of related to this change.

**Comment by nathansobo** [comment]
@jasonrudolph This looks like a great improvement. We make the tabIndex of the editor text box `-1` just to *allow* it to be focused without making an assumption about tab order. As this PR demonstrates, if we're given a clear tabIndex on the parent element there should be no issue (that I know of) forwarding it to the hidden input element.

:zap: Nice work!

File: src/text-editor-element.js [hunk 1]
```diff
@@ -32,7 +32,7 @@ class TextEditorElement extends HTMLElement {
   createdCallback () {
     this.emitter = new Emitter()
     this.initialText = this.textContent
-    this.tabIndex = -1
+    if (this.tabIndex == null) this.tabIndex = -1
```
Comment (nathansobo) [inline src/text-editor-element.js]: 👍

**Comment by jasonrudolph** [comment]
@nathansobo: Thanks for the review! 💟

**Pull request merged**